### PR TITLE
[Storage] Replace Azure CLI with python SDK

### LINKED ISSUE #628: Remove reliance on Libcloud/azcli for Azure blob storage backend
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=37239969)

As we're only using a small subset of Libcloud's features and it's currently limiting our ability to make Medusa faster and safer, we have decided to move away from it and implement async http operations for multipart uploads.

In that effort, we will be moving the Azure storage classes to rely on [the official Azure storage python library](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/storage/azure-storage-blob) that should handle the following features:

- multipart uploads with automated retries of failed parts
- concurrent uploads

This also allows us to remove:

- The reliance on the azcli (and its dependencies)
- The concurrent code that was built to upload multiple files at the same time

This new implementation will use the same global threshold defined for using multipart uploads. 

Implementing rate limiting of uploads is out of scope for this ticket.

## Definition of Done
- [ ] azcli isn't used anymore, nor any command line tool to upload files to Azure Blob Storage
- [ ] Resilient and concurrent multipart uploads are implemented for the Azure backend

### LINKED ISSUE #637: Retry individual blobs download instead of retrying the whole download list
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=38343509)

Currently the retry annotation is placed on `download_blobs()` which results in redownloading all the files when one fails. We need to move the annotation on the `_download_blob()` method instead, so that we only retry failed files individually.

## Definition of Done
- [ ] for all backends, the retry annotation is placed on `_download_blob()` instead of `download_blobs()`
- [ ] gcs present in #631 
- [ ] azure present in #638 
- [x] s3 already present in [here](https://github.com/thelastpickle/cassandra-medusa/blob/master/medusa/storage/s3_base_storage.py#L396)

### PR DESCRIPTION
**Pull request by rzvoncek**
Fixes #628.
Fixes #637.

### CODE REVIEW

**Comment by rzvoncek** [comment]
Rebased on the recent master.

File: medusa/storage/azure_storage.py [hunk 1]
```diff
@@ -1,81 +1,336 @@
...
             )
+        )
 
-        return driver
+        with open(src, "rb") as data:
+            blob_client = await self.azure_container_client.upload_blob(
+                name=object_key,
+                data=data,
+                overwrite=True,
+                max_concurrency=4,
```
Comment (adejanovski) [inline medusa/storage/azure_storage.py:183]: **Todo:** after doing some benchmarks, it looks like 16 gives us better performance. Could you set this value?

File: medusa/storage/azure_storage.py [hunk 2]
```diff
@@ -1,81 +1,336 @@
...
-            multi_part_upload_threshold=int(self.config.multi_part_upload_threshold)
-        )
-
-    def download_blobs(self, srcs, dest):
-        return medusa.storage.azure_blobs_storage.concurrent.download_blobs(
-            self, srcs, dest, self.bucket,
-            max_workers=self.config.concurrent_transfers,
-            multi_part_upload_threshold=int(self.config.multi_part_upload_threshold)
-        )
+    def _get_or_create_event_loop(self) -> asyncio.AbstractEventLoop:
```
Comment (adejanovski) [inline medusa/storage/azure_storage.py:323]: **Suggestion:** this seems to be repeated in all the storage classes, maybe we could refactor it to have a single version that all storage classes will use? In the abstract storage possibly?

**Review by adejanovski — CHANGES_REQUESTED** [review]
One tiny change request and a refactoring ticket suggestion, then we should be good for a merge.

File: medusa/storage/azure_storage.py (hunk 1 above)
Comment (rzvoncek) [reply medusa/storage/azure_storage.py]: Pushed a change.

File: medusa/storage/azure_storage.py (hunk 2 above)
Comment (rzvoncek) [reply medusa/storage/azure_storage.py:323]: It's indeed repeated, I'll be pushing a different PR fixing this.

**Review by adejanovski — APPROVED** [review]

**Pull request merged**